client/hwlib/src/collectors/os_info.rs

```diff
@@ -18,21 +18,36 @@
  *        Nadzeya Hutsko <nadzeya.hutsko@canonical.com>
  */
 
-use anyhow::{anyhow, Context, Result};
+use anyhow::{anyhow, Result};
 use lazy_static::lazy_static;
 use regex::Regex;
 use std::{fs::read_to_string, path::Path, process::Command};
 
 use crate::models::software::{KernelPackage, OS};
 
-impl TryFrom<&Path> for OS {
-    type Error = anyhow::Error;
+pub(crate) trait CommandRunner {
+    fn run_command(&self, cmd: &str, args: &[&str]) -> Result<String>;
+}
+
+pub(crate) struct SystemCommandRunner;
+
+impl CommandRunner for SystemCommandRunner {
+    fn run_command(&self, cmd: &str, args: &[&str]) -> Result<String> {
+        let output = Command::new(cmd).args(args).output()?;
+        let stdout = String::from_utf8(output.stdout)?;
+        Ok(stdout)
+    }
+}
 
-    fn try_from(proc_version_filepath: &Path) -> Result<Self> {
-        let codename = get_codename()?;
-        let distributor = get_distributor()?;
-        let version = get_version()?;
-        let kernel = KernelPackage::try_from(proc_version_filepath)?;
+impl OS {
+    pub(crate) fn try_new<C>(proc_version_filepath: &Path, runner: &C) -> Result<Self>
+    where
+        C: CommandRunner,
```

Comment:
I see a little inconsistency around parameters which implement `CommandRunner` are taken, specifically, we have:
- `fn some_fn<C>(runner: &C) where C: CommandRunner` and
- `fn some_fn(runner: &impl CommandRunner)`

I think we should try to have just one, I think `&impl CommandRunner` is nicer as it doesn't imply that an explicit type parameter is a requirement to define this functions type or implement it. (Also it doesn't require us to specify `C` in three places :P)